### PR TITLE
Bump h5py==3.1.0, bluepy==2.0.0 + bluepysnap/neurom/voxcell/bluepy-configfile

### PR DESCRIPTION
**Pull request by tomdele**
Bump h5py==3.1.0, bluepy==2.0.0, bluepysnap==0.9.0, neurom==1.7.0, voxcell==3.0.0, bluepy-configfile==0.1.15

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-h5py/package.py [hunk 1]
```diff
@@ -17,6 +17,7 @@ class PyH5py(PythonPackage):
     import_modules = ['h5py', 'h5py._hl']
 
     version('master', branch='master')
+    version('3.1.0', sha256='1e2516f190652beedcb8c7acfa1c6fa92d99b42331cbef5e5c7ec2d65b0fc3c2')
```
Comment (genric) [inline var/spack/repos/builtin/packages/py-h5py/package.py:19]: In upstream `spack/develop` there is more stuff related to v3, not sure if we want to ignore those:
`git diff spack/develop -- var/spack/repos/builtin/packages/py-h5py/package.py`

Comment (tomdele) [reply var/spack/repos/builtin/packages/py-h5py/package.py:19]: I was about to ask actually... I am afraid to break everything if I push a new version.

**Comment by tomdele** [comment]
I think I should wait for a spack update with h5py > 3. 
Also I am not quite sure why my build failed due to h5py since I added it and verify all the dependencies.

File: var/spack/repos/builtin/packages/py-h5py/package.py (hunk 1 above)
Comment (matz-e) [reply var/spack/repos/builtin/packages/py-h5py/package.py:19]: I would pull in the remaining changes, too. They seem mostly build-setup stuff, and would probably be good to have.

**Comment by matz-e** [comment]
> I think I should wait for a spack update with h5py > 3.
> Also I am not quite sure why my build failed due to h5py since I added it and verify all the dependencies.

Click on the Jenkins check / "details", then select the "test" tab:

https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.spack-deployment.pull-requests/detail/hpc.spack-deployment.pull-requests/1391/tests

The `configure` step that's conditionally removed upstream fails for all installations of `py-h5py` in the new version.

**Comment by tomdele** [comment]
Thank you @matz-e, I was not familiar with the UI ! It helps a lot indeed.

Gonna push the upstream version of h5py

**Comment by matz-e** [comment]
There are still some pins on h5py that don't account for 3:
```
❯ rg -C1 py-h5py@ var/spack
var/spack/repos/builtin/packages/py-tmd/package.py
22-    depends_on('py-enum34@1.0:', type='run', when='^python@:3.3.99')
23:    depends_on('py-h5py@2.8:2.999', type='run')
24-    depends_on('py-munkres@1.0.12:', type='run')
```
`py-tmd` shows up in the dependency try of `archngv`, along with `py-voxcell`. The following also offend:
```
var/spack/repos/builtin/packages/py-diameter-synthesis/package.py
var/spack/repos/builtin/packages/py-minis-validation/package.py
```
could these 3 packages also accept `py-h5py@3`?

**Comment by tomdele** [comment]
@eleftherioszisis ping :D

**Comment by eleftherioszisis** [comment]
Hi guys! Could we remove that archngv 0.0.0 ? It is a very old version that is not used.

**Comment by tomdele** [comment]
@matz-e is there a special thing to do to remove a package ? Should I just remove it or is it more complex ?

**Comment by matz-e** [comment]
> @matz-e is there a special thing to do to remove a package ? Should I just remove it or is it more complex ?

The whole package? Or just that version?

If you remove `archngv` from `deploy/environments/applications.yaml`, it will stop being deployed. Is this really the desired behavior?

**Comment by tomdele** [comment]
@eleftherioszisis can you confirm you want to remove the whole package and not deploy it anymore ?
(we may recreate a package in the correct directory py-archngv with the new versions)

**Comment by eleftherioszisis** [comment]
I confirm!

**Comment by matz-e** [comment]
@tomdele I would still only remove it from the deployment environment, since deleting the whole package may get us in trouble with the version that is already deployed. Maybe add a depreciation notice comment and comment out the version?

**Comment by tomdele** [comment]
>@tomdele I would still only remove it from the deployment environment, since deleting the whole package may get us in trouble with the version that is already deployed. Maybe add a depreciation notice comment and comment out the version?

sure !

**Comment by matz-e** [comment]
Retest this please

There's some weird bug in the system still…

File: var/spack/repos/builtin/packages/py-h5py/package.py [hunk 2]
```diff
@@ -11,12 +11,12 @@ class PyH5py(PythonPackage):
     HDF5 library from Python."""
 
     homepage = "http://www.h5py.org/"
-    url      = "https://pypi.io/packages/source/h/h5py/h5py-2.10.0.tar.gz"
+    pypi     = "h5py/h5py-3.1.0.tar.gz"
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/py-h5py/package.py:14]:
```suggestion
    pypi     = "h5py/h5py-3.1.0.tar.gz"
    url      = "https://pypi.io/packages/source/h/h5py/h5py-2.10.0.tar.gz"
```
This `pypi` is only on the newest Spack and didn't make it into the release we're using. Adding the URL should be fine, will be easy to merge next time around :)

**Review by matz-e — CHANGES_REQUESTED** [review]

**Comment by tomdele** [comment]
Could it be the not having any version number for archngv messing up everything ?  

`Error: unorderable types: str() < NoneType()`
Make me think about something like that...

**Comment by matz-e** [comment]
Retest this please

> Could it be the not having any version number for archngv messing up everything ?
> 
> `Error: unorderable types: str() < NoneType()`
> Make me think about something like that...

That would show up in `applications`, IIRC. This has to do with virtual dependencies like `mpi` that can be fulfilled by more than one package. I have not yet found out what triggers it, and I tend to have a hard time getting it to crash like that myself :(

**Comment by tomdele** [comment]
What about keeping archngv but adding all the version switches for the version of h5py<3.0.0 ? would it be easier ?

**Comment by matz-e** [comment]
> What about keeping archngv but adding all the version switches for the version of h5py<3.0.0 ? would it be easier ?

Sorry, took me a little to dig into this. I think that `py-diameter-synthesis` still pins h5py to a lower version. The error/traceback actually comes from `placement-algorithm`…

**Comment by eleftherioszisis** [comment]
@arnaudon , @adrien-berchet , @wizmer are involved with the tools Matthias mentioned.

**Comment by arnaudon** [comment]
Hello, what do you need us to do on these 2 packages to be able to resolve your issue? We are flexible on h5py version, etc...

**Comment by matz-e** [comment]
> Hello, what do you need us to do on these 2 packages to be able to resolve your issue? We are flexible on h5py version, etc...

If `diameter-synthesis` can handle `h5py` version 3, then that version pin can be removed / opened up.

**Comment by tomdele** [comment]
> Sorry, took me a little to dig into this.

no worries ! I am so glad you can help me on this one ...

**Comment by adrien-berchet** [comment]
> If `diameter-synthesis` can handle `h5py` version 3, then that version pin can be removed / opened up.

I see no reason it could not handle it. Actually, this requirements seems useless in `diameter-synthesis`, I will remove it.

**Comment by matz-e** [comment]
OK, I've fiddled around with h5py enough to get it to concretize the dependency try locally. I hope that Jenkins will run through successfully now, but the results should be checked, since this is pretty invasive…

**Comment by matz-e** [comment]
Retest this please

And we still have the heisenbug in the CI…

**Comment by matz-e** [comment]
Retest this please

And again… sorry for the noise

**Comment by matz-e** [comment]
Retest this please

Now it at least looks like just a race condition

**Comment by tomdele** [comment]
> And again… sorry for the noise

No worries. Is there anything I can do to help ? I feel helpless.

**Comment by matz-e** [comment]
> > And again… sorry for the noise
> 
> No worries. Is there anything I can do to help ? I feel helpless.

This looks like an issue that may have arisen by a previous PR to deal with some Intel compiler stuff. It's ugly and probably more of an HPC thing.

I'm a bit concerned about all the version modifications that I did. As long as you/the rest of NSE is OK with that, I'll try to get that compilation of scipy fixed.

**Comment by matz-e** [comment]
We will be able to proceed with this once I've done https://bbpteam.epfl.ch/project/issues/browse/BSD-208

**Comment by tomdele** [comment]
> I'm a bit concerned about all the version modifications that I did. As long as you/the rest of NSE is OK with that, I'll try to get that compilation of scipy fixed.

I am perfectly fine with the changes you did and anyway they will need to be done at some point. This is a breaking change in one of our main backend (h5py) so i cannot expect changing this without a risk of breakage ...

**Comment by genric** [comment]
Any estimate on this?

**Comment by matz-e** [comment]
> Any estimate on this?

I got scipy to build with the latest fixes in #1016. Just took me a little extra time since `zsh` is not as "special" as `bash`…

**Comment by matz-e** [comment]
Try rebasing this, please?

**Comment by tomdele** [comment]
> Try rebasing this, please?

Done. I hope I did not screw this up ...

**Review by matz-e — COMMENTED** [review]
```
File "/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1007/deploy/applications/2021-01-14/linux-rhel7-x86_64/gcc-9.3.0/py-neurom-1.7.0-txvfbn/lib/python3.8/site-packages/neurom/io/hdf5.py", line 45, in <module>

    import h5py

  File "/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1007/deploy/applications/2021-01-14/linux-rhel7-x86_64/gcc-9.3.0/py-h5py-3.1.0-6v5agu/lib/python3.8/site-packages/h5py/__init__.py", line 60, in <module>

    from ._hl.files import (

  File "/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1007/deploy/applications/2021-01-14/linux-rhel7-x86_64/gcc-9.3.0/py-h5py-3.1.0-6v5agu/lib/python3.8/site-packages/h5py/_hl/files.py", line 35, in <module>

    libver_dict.update({'v108': h5f.LIBVER_V18, 'v110': h5f.LIBVER_V110})

AttributeError: module 'h5py.h5f' has no attribute 'LIBVER_V18'
```
sigh :(

**Comment by matz-e** [comment]
Also this:
```
/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1007/deploy/applications/2021-01-14/linux-rhel7-x86_64/gcc-9.3.0/py-h5py-3.1.0-6v5agu/lib/python3.8/site-packages/h5py/__init__.py:36: UserWarning: h5py is running against HDF5 1.10.7 when it was built against 1.8.12, this may cause problems
```

**Comment by tomdele** [comment]
omg ... 

So if I understand correctly : 
we are using  1.8.12 to build h5py but running it with 1.10.7 and so in line :
https://github.com/h5py/h5py/blob/master/h5py/_hl/files.py#L34
hdf5_version is 1.10.7 and we are trying to get `{h5f.LIBVER_V18: 'v108', h5f.LIBVER_V110: 'v110'}`
that does not exists due to HDF5_VERSION being 1.8.12 in:
https://github.com/h5py/h5py/blob/master/h5py/h5f.pyx#L64

**Comment by matz-e** [comment]
> omg ...
> 
> So if I understand correctly :
> we are using 1.8.12 to build h5py but running it with 1.10.7 and so in line :
> https://github.com/h5py/h5py/blob/master/h5py/_hl/files.py#L34
> hdf5_version is 1.10.7 and we are trying to get `{h5f.LIBVER_V18: 'v108', h5f.LIBVER_V110: 'v110'}`
> that does not exists due to HDF5_VERSION being 1.8.12 in:
> https://github.com/h5py/h5py/blob/master/h5py/h5f.pyx#L64

Sorry, was poking around a little more. We should be using 1.10.7 consistently, even when building

This may just be due to the additional glue to build h5py. To be sure, I'm nuking the PR build directory on BB5, and we'll start building this PR from scratch.

**Comment by tomdele** [comment]
Ok let me know.

**Comment by matz-e** [comment]
Retest this please

**Comment by matz-e** [comment]
I have the suspicion that somehow, this picks up the system HDF5 rather than our Spack one

**Comment by matz-e** [comment]
Reproduced :(

With just `spack install py-h5py`, we have an implicit `+mpi`, and this issue does not pop up. but explicitly requiring `~mpi`, I also see it picking up the wrong HDF5 <insert old man shaking fist at cloud.gif>

**Comment by tomdele** [comment]
> py-h5py: bring in line with upstream 

Very unlucky ... the change in upstream has been made the same day as me updating the py-h5py package ...

**Comment by matz-e** [comment]
> > py-h5py: bring in line with upstream
> 
> Very unlucky ... the change in upstream has been made the same day as me updating the py-h5py package ...

Yeah, I was looking at that. Bad coincidence, but at least for me this was reasonably quick to notice the fix. This `%intel` with `scipy` on the other hand :(

**Comment by tomdele** [comment]
> This %intel with scipy on the other hand :(

Yes, I can imagine ... hardware specific compiler problems for a specific package ...

File: var/spack/repos/builtin/packages/py-bluepy/package.py [hunk 3]
```diff
@@ -28,26 +29,51 @@ class PyBluepy(PythonPackage):
...
+    depends_on('py-libsonata@0.1.3:0.1.4', type='run', when='^python@:3.5.99')
+
     depends_on('brion+python@3.1.0:', type='run')
+    depends_on('py-bluepy-configfile@0.1.11:', type='run')
+    depends_on('py-numpy@1.8:', type='run')
+
+    # for old versions of bluepy <2.0.0
+    depends_on('py-enum34@1.0:', type='run', when='^python@:3.3.99')
+    depends_on('py-lxml@3.3:', type='run', when='@:1.0.0')
+    depends_on('py-pylru@1.0:', type='run', when='@:1.0.0')
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/py-bluepy/package.py]: Really?
```
  File "/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/pulls/1007/deploy/applications/2021-01-27/linux-rhel7-x86_64/gcc-9.3.0/py-bluepy-2.0.0-o5qez4/lib/python3.8/site-packages/bluepy/morphology.py", line 11, in <module>
    from pylru import FunctionCacheManager
ModuleNotFoundError: No module named 'pylru'
```
Let's hope this is the last thing :)

Comment (tomdele) [reply var/spack/repos/builtin/packages/py-bluepy/package.py]: My bad on this one...

**Comment by tomdele** [comment]
And now it s brainbuilder turn to fail \o/ ...

**Comment by matz-e** [comment]
> And now it s brainbuilder turn to fail \o/ ...

Everybody wants a turn ;)

**Comment by tomdele** [comment]
I will fix this this morning.

**Comment by genric** [comment]
:fireworks: , no?

**Review by matz-e — APPROVED** [review]

**Comment by tomdele** [comment]
Thank you so much for all the support @matz-e

**Comment by matz-e** [comment]
You're welcome! Glad this one is merged, and it revealed some bugs, too

**Comment by genric** [comment]
The merge didn't trigger the `unstable` deploy, right?

**Comment by matz-e** [comment]
Grrr

**Comment by matz-e** [comment]
I triggered it.

**Pull request merged**